### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by KalleMarSolita**
Functionalities to be tested:
*Revert
*Import
*Fill default values for row/column/columns
*Add localization to input fields
Press "Save" after each test and UI/current localizations should be updated

### CODE REVIEW

File: Solita.LocalizationEditor.UI/DAL/BlobFileAccessStrategy.cs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        {
+            var blobFactory = ServiceLocator.Current.GetInstance<BlobFactory>();
+            var mediaDataResolver = ServiceLocator.Current.GetInstance<ContentMediaResolver>();
+            var contentTypeRepository = ServiceLocator.Current.GetInstance<IContentTypeRepository>();
+
+            var file = ContentRepo.GetDefault<LocalizationFile>(LocalizationStorageReference);
+            file.Name = LocalizationFileName;
+
+            var blob = blobFactory.CreateBlob(file.BinaryDataContainer, ".xml");
+            var xmlBytes = Encoding.UTF8.GetBytes(xml.OuterXml);
```
Comment (jukkahyv) [inline Solita.LocalizationEditor.UI/DAL/BlobFileAccessStrategy.cs]: This line seems to be unused. Somehow I'm thinking there could be a more efficient way of saving the XML anyway, using streams instead of reading and parsing the XML as string.

File: Solita.LocalizationEditor.UI/DAL/VppFileAccessStrategy.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+using System.Web;
+using System.Web.Hosting;
+using System.Xml;
+using EPiServer.Web.Hosting;
+using Solita.LocalizationEditor.UI.Common;
+using Solita.LocalizationEditor.UI.Models;
+
+namespace Solita.LocalizationEditor.UI.DAL
+{
+    //public class VppFileAccessStrategy : FileAccessStrategy
```
Comment (jukkahyv) [inline Solita.LocalizationEditor.UI/DAL/VppFileAccessStrategy.cs]: I assume this is commented out because it's not compatible with EPi 9? Should it be moved to a separate library?

File: Solita.LocalizationEditor.UI/Helpers/XmlLanguageFileHelper.cs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        }
+
+        public void AddLanguageNames(XmlDocument xmlDoc)
+        {
+            var nodes = xmlDoc.SelectNodes(LanguageXPath);
+            if (nodes == null) { return; }
+
+            foreach (XmlNode node in nodes)
+            {
+                var id = node.Attributes["id"].InnerText;
```
Comment (jukkahyv) [inline Solita.LocalizationEditor.UI/Helpers/XmlLanguageFileHelper.cs]: There's some duplication here...

File: Solita.LocalizationEditor.UI/LocalizationPersister.cs [hunk 4]
```diff
@@ -61,130 +61,140 @@ public CategoryList GetLocalizations()
...
 
-            var localizations = LocalizationHelpers.GetLocalizationDefinitions().Select(d => new
-                {
-                    key = d.Key,
-                    translations = languages.Select(language => new {language, value = FindExistingTranslation(xml, language, d.Key)})
-                });
+            var matchingDefinitions =
+                    from definition in definitions
+                    let translations = FindExistingTranslations(xml, languages, definition.Key)
+                    where translations.Count() > 0
```
Comment (jukkahyv) [inline Solita.LocalizationEditor.UI/LocalizationPersister.cs]: if you're using Count(), might as well use .Any()

File: Solita.LocalizationEditor.UI/Models/CategoryList.cs [hunk 5]
```diff
@@ -47,6 +48,36 @@ public void AddTranslation(string languageId, string value)
...
+
+            public override bool Equals(object obj)
+            {
+                var translation = obj as Translation;
+                if (translation == null)
+                {
+                    return false;
+                }
+                if (
+                    string.Compare(translation.Key, Key) != 0 ||
```
Comment (jukkahyv) [inline Solita.LocalizationEditor.UI/Models/CategoryList.cs]: Just wondering why you didn't use string.Equals here since that seems more appropriate

File: Solita.LocalizationEditor.UI/Modules/_protected/Solita.LocalizationEditor.UI/Views/LocalizationEditor/Index.cshtml [hunk 6]
```diff
@@ -33,7 +34,7 @@
         <div id="epi-globalDocument">
             @Html.Raw(Html.GlobalMenu())
 
-            <div class="localization-editor" data-localizations-url="@Url.Action("GetJsonLocalizations")">
+            <div class="localization-editor" data-localizations-url="@Url.Action("GetJsonLocalizations")" data-localizations-transform-url="@Url.Action("TransformLocalizationXmlToJsonLocazlitaions")">
```
Comment (jukkahyv) [inline Solita.LocalizationEditor.UI/Modules/_protected/Solita.LocalizationEditor.UI/Views/LocalizationEditor/Index.cshtml]: Locazlitaions?

File: Solita.LocalizationEditor.UI/Properties/AssemblyInfo.cs [hunk 7]
```diff
@@ -6,7 +6,7 @@
 // set of attributes. Change these attribute values to modify the information
 // associated with an assembly.
 [assembly: AssemblyTitle("Solita.LocalizationEditor")]
-[assembly: AssemblyDescription("Localization Editor for EPiServer CMS")]
+[assembly: AssemblyDescription("LocalizationResult Editor for EPiServer CMS")]
```
Comment (jukkahyv) [inline Solita.LocalizationEditor.UI/Properties/AssemblyInfo.cs]: Was this change intentional?

File: Solita.LocalizationEditor.UI/DAL/VppFileAccessStrategy.cs (hunk 2 above)
Comment (KalleMarSolita) [inline Solita.LocalizationEditor.UI/DAL/VppFileAccessStrategy.cs]: This class was removed in latest commit as redundant.

File: Solita.LocalizationEditor.UI/Properties/AssemblyInfo.cs (hunk 7 above)
Comment (KalleMarSolita) [inline Solita.LocalizationEditor.UI/Properties/AssemblyInfo.cs]: Very good point, no it wasn't. Accidentally performed replace string for this one and didn't notice it...

File: Solita.LocalizationEditor.UI/DAL/BlobFileAccessStrategy.cs [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+            var mediaData = GetLocalizationFiles().FirstOrDefault();
+
+            return mediaData ?? ContentRepo.GetDefault<LocalizationFile>(LocalizationStorageReference);
+        }
+
+        public override void SaveXml(XmlDocument xml)
+        {
+            var blobFactory = ServiceLocator.Current.GetInstance<BlobFactory>();
+            var mediaDataResolver = ServiceLocator.Current.GetInstance<ContentMediaResolver>();
+            var contentTypeRepository = ServiceLocator.Current.GetInstance<IContentTypeRepository>();
```
Comment (tommiseppanen) [inline Solita.LocalizationEditor.UI/DAL/BlobFileAccessStrategy.cs]: mediaDataResolver & contentTypeRepository are never used (in addition to the xmlBytes)

File: Solita.LocalizationEditor.UI/DAL/FileAccessStrategy.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
+﻿using System.Collections.Generic;
+using System.Xml;
+using EPiServer.Framework.Blobs;
+using Solita.LocalizationEditor.UI.Models;
+
+namespace Solita.LocalizationEditor.UI.DAL
+{
+    public abstract class FileAccessStrategy
+    {
+        public string FilePath { get; }
```
Comment (tommiseppanen) [inline Solita.LocalizationEditor.UI/DAL/FileAccessStrategy.cs]: It looks that FilePath property is never used. By the way, did you consider using interface instead of abstract class for FileAccessStrategy?

File: Solita.LocalizationEditor.UI/Business/BlobLocalizationProviderInitiator.cs [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+
+        public void ReInitProvider()
+        {
+            var service = ServiceLocator.Current.GetInstance<LocalizationService>() as ProviderBasedLocalizationService;
+            if (service == null)
+            {
+                Log.Error("LocalizationService unavailable");
+                return;
+            }
+            var translationProvider = this.GetTranslationProvider();
```
Comment (tommiseppanen) [inline Solita.LocalizationEditor.UI/Business/BlobLocalizationProviderInitiator.cs]: "this" keyword is used inconsistently. It is redundant in all of the cases in this file and could be removed.

File: Solita.LocalizationEditor.UI/Business/BlobLocalizationProviderInitiator.cs [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+
+    [InitializableModule]
+    //[ModuleDependency(typeof(FrameworkInitialization))]
+    [ModuleDependency(typeof(EPiServer.Web.InitializationModule))] 
+    public class BlobLocalizationProviderInitiator : IInitializableModule
+    {
+        private const string ProviderName = "Translations";
+        private static readonly ILog Log = LogManager.GetLogger(typeof(BlobLocalizationProviderInitiator));
+        private static bool initialized;
+        private ProviderBasedLocalizationService localizationService;
```
Comment (tommiseppanen) [inline Solita.LocalizationEditor.UI/Business/BlobLocalizationProviderInitiator.cs]: Couple of coding style notes: The commented code line (ModuleDependency) should be removed and the suggested names for "initialized" & "localizationService" would be "_initialized" & "_localizationService"
